packages/core/src/entity/EntityStoreBase.ts

```diff
@@ -27,6 +30,120 @@ export class EntityStoreBase {
+        if (!entityToResolveFieldsFrom) {
+          throw new Error(
+            `Logic Error: Cannot resolve field ${field} of a fieldset as there is no entity to resolve it from.`,
+          );
+        }
+        if (isLeaf) {
+          resolvedFieldset.push([entityToResolveFieldsFrom, field, _localeQualifier]);
+          break;
+        }
+
+        const fieldValue = get<string | UnresolvedLink<'Entry' | 'Asset'>>(
+          entityToResolveFieldsFrom,
+          ['fields', field],
+        );
+
+        if (undefined === fieldValue) {
+          return {
+            resolvedFieldset,
+            isFullyResolved: false,
+            reason: `Cannot resolve field Link<${entityToResolveFieldsFrom.sys.type}>(sys.id=${entityToResolveFieldsFrom.sys.id}).fields[${field}] as field value is not defined`,
+          };
+        } else if (isLink(fieldValue)) {
+          const entity = this.getEntityFromLink(fieldValue);
+          if (entity === undefined) {
+            return {
+              resolvedFieldset,
+              isFullyResolved: false,
+              reason: `Field reference Link (sys.id=${fieldValue.sys.id}) not found in the EntityStore, waiting...`,
+            };
+          }
+          resolvedFieldset.push([entityToResolveFieldsFrom, field, _localeQualifier]);
+          entityToResolveFieldsFrom = entity; // we move up
+        } else {
+          // TODO: Eg. when someone changed the schema and the field is not a link anymore, what should we return then?
+          throw new Error(
+            `LogicError: Invalid value of a field we consider a reference field. Cannot resolve field ${field} of a fieldset as it is not a link, neither undefined.`,
+          );
+        }
+      }
+      return {
+        resolvedFieldset,
+        isFullyResolved: true,
+      };
+    };
+
+    if (isDeepPath(path)) {
+      const headEntity = isLink(linkOrEntryOrAsset)
+        ? this.getEntityFromLink(linkOrEntryOrAsset)
+        : (linkOrEntryOrAsset as Entry | Asset);
+
+      if (undefined === headEntity) {
+        return;
+      }
+      const unresolvedFieldset = parseDataSourcePathIntoFieldset(path);
+
+      // The purpose here is to take this intermediate representation of the deep-path
+      // and to follow the links to the leaf-entity and field
+      // in case we can't follow till the end, we should signal that there was null-reference in the path
+      const { resolvedFieldset, isFullyResolved, reason } = resolveFieldset(
+        unresolvedFieldset,
+        headEntity,
+      );
+      if (!isFullyResolved) {
+        reason &&
+          console.debug(
+            `[exp-builder.sdk::EntityStoreBased::getValueDeep()] Deep path wasn't resolved till leaf node, falling back to undefined, because: ${reason}`,
+          );
+        return undefined;
+      }
+      const [leafEntity] = resolvedFieldset[resolvedFieldset.length - 1];
+      return leafEntity;
+    } else {
```

Comment:
I think removing this `else` statement could simplify the flow and slightly reduce indentation. Since the `if` condition above early returns, the `else` clause can be avoided.

Reply:
Normally I'd agree, but there was quite a bit of logic in both statements and I think it becomes confusing in that case. So to combat that, I refactored out the code in the top if block into a new `getDeepEntry` method on the class and was able to do this.